cinder/volume/drivers/nexenta/ns5/nfs.py

```diff
@@ -265,14 +382,38 @@ def delete_volume(self, volume):
             [pool, fs, volume['name']])
 
         field = 'originalSnapshot'
-        origin = self.nef.get('{}?fields={}'.format(url, field)).get(field)
+        origin = self.nef.get(url).get(field)
+        url = 'storage/filesystems/%s?snapshots=true' % '%2F'.join(
+            [pool, fs, volume['name']])
         try:
             self.nef.delete(url)
         except exception.NexentaException as exc:
-            vol_path = '/'.join((self.share, volume['name']))
-            self.destroy_later_or_raise(exc, vol_path)
-            return
-        self.collect_zfs_garbage(origin)
+            if 'Failed to destroy snap' in exc.kwargs['message']['message']:
+                url = 'storage/snapshots?parent=%s' % '%2F'.join(
+                    [pool, fs, volume['name']])
+                snap_list = []
+                snap_map = {}
+                for snap in self.nef.get(url)['data']:
```

Comment:
The same about combining two loops: snap_list is used only to cycle through it in the next loop

Reply:
fixed